llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,386 @@
+
+  if (LoopMBBs.empty())
+    return false;
+
+  this->MF = &MF;
+  MRI = &MF.getRegInfo();
+  TRI = static_cast<const AIEBaseRegisterInfo *>(MRI->getTargetRegisterInfo());
+  VRM = &getAnalysis<VirtRegMap>();
+  LRM = &getAnalysis<LiveRegMatrix>();
+  LIS = &getAnalysis<LiveIntervals>();
+  bool Modified = false;
+
+  LLVM_DEBUG(dbgs() << "*** WAW Loop Register Rewriting: " << MF.getName());
+  LLVM_DEBUG(dbgs() << " ***\n");
+
+  for (const MachineBasicBlock *MBB : LoopMBBs)
+    Modified |= renameMBBPhysRegs(MBB);
+
+  return Modified;
+}
+
+/// return a vector containing virtual registers, that have had a copy to define
+/// the virtual register
+SmallVector<Register, 16> getVRegWithCopies(const MachineBasicBlock &MBB) {
+  SmallVector<Register, 16> VRegWithCopies;
+  for (const MachineInstr &MI : MBB) {
+    for (const MachineOperand Def : MI.defs()) {
+      Register Reg = Def.getReg();
+
+      if (!Reg.isVirtual())
+        continue;
+
+      if (MI.isCopyLike() && find(VRegWithCopies, Reg) == VRegWithCopies.end())
+        VRegWithCopies.push_back(Reg);
+    }
+  }
+
+  return VRegWithCopies;
+}
+
+bool AIEWawRegRewriter::renameMBBPhysRegs(const MachineBasicBlock *MBB) {
+  LLVM_DEBUG(dbgs() << "WAW Reg Renaming BasicBlock "; MBB->dump();
+             dbgs() << "\n");
+  bool Modified = false;
+  // Add the used physical registers one machine instruction at a time.
+  // This vector is used to determine, if a physical register has already been
+  // defined in the machine basic block.
+  BitVector UsedPhysRegs(TRI->getNumRegs());
+
+  // Get a list of registers, that are not allowed as a replacement register.
+  // This list gets updated with the newly replaced physical register, so that
+  // this pass does not introduce WAW dependencies.
+  BitVector BlockedPhysRegs = getDefinedPhysRegs(MBB);
+
+  // Collect all the virtual registers that have at least a copy instruction
+  // that defines them. Subregisters may contain constants that may be shared
+  // across different virtual registers. Renaming would reintroduce unnecessary
+  // copies, if physical registers are shared. Also do not rename copies, since
+  // they could be removed in a later pass.
+  SmallVector<Register, 16> VRegWithCopies = getVRegWithCopies(*MBB);
+
+  std::map<Register, const MachineInstr *> LastVRegDef = getLastVRegDef(*MBB);
+
+  for (const MachineInstr &MI : *MBB) {
+
+    // Identity copies will be removed in a later pass, therefore, these are not
+    // real defines of a physical register
+    if (isIdentityCopy(MI))
+      continue;
+
+    for (const MachineOperand &MO : MI.defs()) {
+
+      Register Reg = MO.getReg();
+      if (!Reg.isVirtual())
+        continue;
+      if (VRM->hasRequiredPhys(Reg))
+        continue;
+      if (MO.isTied())
+        continue;
+      // wait until the virtual register is fully defined to check if it is a
+      // replacement candidate
+      if (LastVRegDef[Reg] != &MI)
```

Comment:
I think the comment could be that we don't care about several definitions of the same virtual register because even if we were to rename the virtual register, by construction we would rename all the definition and achieving nothing wrt WAW :).
I.e., it's not just about fully defined.

Reply:
updated the comment